polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -164,6 +175,92 @@ static bool isClassType(Type type, const llvm::Regex &regex) {
   return st && st.isIdentified() && regex.match(st.getName());
 }
 
+/// Check \p op is a member of class \p className and return the first (this)
+/// argument.
+static Value getThisArgument(FunctionOpInterface op, StringRef className) {
+  llvm::ItaniumPartialDemangler demangler;
+  if (!(succeeded(partialDemangle(demangler, op)) && demangler.isFunction() &&
+        isMemberFunction(demangler, syclNamespace, className)))
+    return nullptr;
+
+  assert(op.getNumArguments() > 0 &&
+         "A non-static member function must have at least one argument");
+  Value firstArg = op.getArgument(0);
+  assert(isa<LLVM::LLVMPointerType>(firstArg.getType()) &&
+         "Expecting this argument to be a pointer");
+  return firstArg;
+}
+
+static FailureOr<StringRef> getStringValue(LLVM::GlobalOp op) {
+  // Check the operation has a value
+  std::optional<Attribute> attr = op.getValue();
+  if (!attr)
+    return failure();
+
+  // Check it is a string
+  auto strAttr = dyn_cast<StringAttr>(*attr);
+  if (!strAttr)
+    return failure();
+
+  // Drop the trailing `0` character
+  return strAttr.getValue().drop_back();
+}
+
+static FailureOr<StringRef> getAnnotation(LLVM::VarAnnotation annotation) {
+  auto addressofOp =
+      annotation.getAnnotation().getDefiningOp<LLVM::AddressOfOp>();
+  if (!addressofOp)
+    return failure();
+
+  SymbolTableCollection symbolTable;
+  auto global = symbolTable.lookupNearestSymbolFrom<LLVM::GlobalOp>(
+      addressofOp, addressofOp.getGlobalNameAttr());
+  return global ? getStringValue(global) : FailureOr<StringRef>(failure());
+}
+
+/// Return the constructor reponsible of the construction of \p value.
```

Comment:
Typo: responsible
Maybe document here that this method returns `nullptr` if interfering users are found.

Reply:
Done